app/models/badge_rule.rb

```diff
@@ -35,8 +36,8 @@ def rule_satisfied?(test_passage)
   private
 
   def user_completed_all_backend_tests?(user)
-    backend_tests = Test.joins(:category).where(categories: { title: 'Backend' })
-    completed_tests = user.test_passages.passed.pluck(:test_id).uniq
+    backend_tests = Test.by_category('Backend')
+    completed_tests = user.test_passages.passed.map(&:test_id).last(1)
```

Comment:
Are you sure that we need redo this completed_tests from uniq array of id tp last id?
Because looks like problem with logic)))

anyway, even this implementation, why you do not do like
user.test_passages.passed.last.test_id
in this case you won't fetch each object, then get test_id from each only for to get the last one)